File: src/main/java/io/gittul/domain/user/UserService.java
package io.gittul.domain.user;

import io.gittul.domain.tag.TagService;
import io.gittul.domain.tag.entity.Tag;
import io.gittul.domain.tag.entity.UserInterest;
import io.gittul.domain.user.entity.User;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;
    private final TagService tagService;

    @Transactional
    public void interest(User user, List<String> tagNames) {
        List<UserInterest> interests = user.getDetails().getInterests();

        List<Tag> tags = tagService.getOrCreate(tagNames);

        List<UserInterest> newInterests = tags.stream()
                .map(tag -> new UserInterest(user, tag))
                .filter(interest -> !interests.contains(interest))
                .toList();

        interests.addAll(newInterests);
        userRepository.save(user);
    }

    public void unInterest(User user, String tagName) {
        user.getDetails().getInterests().removeIf(
                userInterest -> userInterest.getTag().getTagName().equals(tagName)
        );
        userRepository.save(user);
    }
}

File: app/src/main/kotlin/io/gittul/app/domain/user/UserService.kt
package io.gittul.app.domain.user

import io.gittul.app.domain.tag.TagService
import io.gittul.core.domain.tag.entity.Tag
import io.gittul.core.domain.tag.entity.UserInterest
import io.gittul.core.domain.user.entity.User
import jakarta.transaction.Transactional
import org.springframework.stereotype.Service

@Service
class UserService(
    private val userRepository: UserRepository,
    private val tagService: TagService
) {

    @Transactional
    fun interest(
        user: User,
        tagNames: List<String>
    ) {
        val interests: MutableList<UserInterest> = user.details.interests

        val tags: List<Tag> = tagService.getOrCreate(tagNames)

        val newInterests: List<UserInterest> = tags.stream()
            .map<UserInterest> { UserInterest(user, it) }
            .filter { interests.contains(it) }
            .toList()

        interests.addAll(newInterests)
        userRepository.save(user)
    }

    fun unInterest(user: User, tagName: String) {
        user.details.interests
            .removeIf { it.tag.tagName == tagName }
        userRepository.save(user)
    }
}
